Thread JavaFX Application Thread Exception when running save

Thread[JavaFX Application Thread,5,main] threw an exception
## Stack Trace:
```java
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at javafx.fxml.FXMLLoader$MethodHandler.invoke(FXMLLoader.java:1774)
	at javafx.fxml.FXMLLoader$ControllerMethodEventHandler.handle(FXMLLoader.java:1657)
	at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:86)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:49)
	at javafx.event.Event.fireEvent(Event.java:198)
	at javafx.scene.control.MenuItem.fire(MenuItem.java:462)
	at com.sun.javafx.scene.control.skin.ContextMenuContent$MenuItemContainer.doSelect(ContextMenuContent.java:1405)
	at com.sun.javafx.scene.control.skin.ContextMenuContent$MenuItemContainer.lambda$createChildren$343(ContextMenuContent.java:1358)
	at com.sun.javafx.event.CompositeEventHandler$NormalEventHandlerRecord.handleBubblingEvent(CompositeEventHandler.java:218)
	at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:80)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
	at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
	at javafx.event.Event.fireEvent(Event.java:198)
	at javafx.scene.Scene$MouseHandler.process(Scene.java:3757)
	at javafx.scene.Scene$MouseHandler.access$1500(Scene.java:3485)
	at javafx.scene.Scene.impl_processMouseEvent(Scene.java:1762)
	at javafx.scene.Scene$ScenePeerListener.mouseEvent(Scene.java:2494)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:352)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:275)
	at java.security.AccessController.doPrivileged(Native Method)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleMouseEvent$300(GlassViewEventHandler.java:388)
	at com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:389)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.handleMouseEvent(GlassViewEventHandler.java:387)
	at com.sun.glass.ui.View.handleMouseEvent(View.java:555)
	at com.sun.glass.ui.View.notifyMouse(View.java:937)
	at com.sun.glass.ui.gtk.GtkApplication._runLoop(Native Method)
	at com.sun.glass.ui.gtk.GtkApplication.lambda$null$450(GtkApplication.java:139)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sun.reflect.misc.Trampoline.invoke(MethodUtil.java:71)
	at sun.reflect.GeneratedMethodAccessor1.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sun.reflect.misc.MethodUtil.invoke(MethodUtil.java:275)
	at javafx.fxml.FXMLLoader$MethodHandler.invoke(FXMLLoader.java:1771)
	... 43 more
Caused by: java.io.FileNotFoundException: /opt/GRIP/app/Test (Permission denied)
	at java.io.FileOutputStream.open0(Native Method)
	at java.io.FileOutputStream.open(FileOutputStream.java:270)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:213)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:162)
	at edu.wpi.grip.core.serialization.Project.save(Project.java:127)
	at edu.wpi.grip.ui.MainWindowController.saveProjectAs(MainWindowController.java:252)
	... 53 more

```
## System Info:

Property Name | Property
 ----- | -----
GRIP Version | 1.5.1-1-gc311c8e
java.version | 1.8.0_111
javafx.version | 8.0.60
os.name | Linux
os.version | 4.4.0-31-generic
os.arch | amd64

--- a/core/src/main/java/edu/wpi/grip/core/serialization/Project.java
+++ b/core/src/main/java/edu/wpi/grip/core/serialization/Project.java
@@ -3,8 +3,10 @@ package edu.wpi.grip.core.serialization;
 import edu.wpi.grip.core.Pipeline;
 import edu.wpi.grip.core.PipelineRunner;
 import edu.wpi.grip.core.events.DirtiesSaveEvent;
+import edu.wpi.grip.core.events.WarningEvent;
 
 import com.google.common.annotations.VisibleForTesting;
+import com.google.common.eventbus.EventBus;
 import com.google.common.eventbus.Subscribe;
 import com.google.common.reflect.ClassPath;
 import com.thoughtworks.xstream.XStream;
@@ -35,6 +37,8 @@ import javax.inject.Singleton;
 public class Project {
 
   protected final XStream xstream = new XStream(new PureJavaReflectionProvider());
+  @Inject
+  private EventBus eventBus;
   @Inject
   private Pipeline pipeline;
   @Inject
@@ -118,6 +122,29 @@ public class Project {
     this.xstream.fromXML(reader);
     pipelineRunner.startAsync();
     saveIsDirty.set(false);
+  }
+
+  /**
+   * Tries to save this project to the given file. Unlike {@link #save(File)}, this will <i>not</i>
+   * throw an IOException and will instead post a warning event to the event bus.
+   *
+   * @param file the file to save to
+   *
+   * @return true if the project was successfully saved to the given file, or false if the file
+   *              could not be written to
+   */
+  public boolean trySave(File file) {
+    try {
+      save(file);
+      return true;
+    } catch (IOException e) {
+      eventBus.post(new WarningEvent(
+          "Could not save project",
+          "The project could not be saved to " + file.getAbsolutePath()
+              + ".\n\nCause: " + e.getMessage()
+      ));
+      return false;
+    }
   }
 
   /**

--- a/ui/src/main/java/edu/wpi/grip/ui/MainWindowController.java
+++ b/ui/src/main/java/edu/wpi/grip/ui/MainWindowController.java
@@ -34,7 +34,6 @@ import java.io.File;
 import java.io.IOException;
 import java.util.Optional;
 import java.util.Set;
-import java.util.logging.Logger;
 
 import javafx.application.Platform;
 import javafx.fxml.FXML;
@@ -62,8 +61,6 @@ import javax.inject.Inject;
  * The Controller for the application window.
  */
 public class MainWindowController {
-
-  private static final Logger logger = Logger.getLogger(MainWindowController.class.getName());
 
   @FXML
   private Parent root;
@@ -218,16 +215,7 @@ public class MainWindowController {
   public boolean saveProject() {
     if (project.getFile().isPresent()) {
       // Immediately save the project to whatever file it was loaded from or last saved to.
-      try {
-        project.save(project.getFile().get());
-      } catch (IOException e) {
-        eventBus.post(new WarningEvent("Could not save project",
-            "The project could not be saved to " + project.getFile().get().getAbsolutePath()
-                + ".\n\nCause: " + e.getMessage()
-        ));
-        return false;
-      }
-      return true;
+      return project.trySave(project.getFile().get());
     } else {
       return saveProjectAs();
     }
@@ -254,15 +242,7 @@ public class MainWindowController {
       return false;
     }
 
-    try {
-      project.save(file);
-    } catch (IOException e) {
-      eventBus.post(new WarningEvent("Could not save project",
-          "The project could not be saved to " + file.getAbsolutePath()
-              + ".\n\nCause: " + e.getMessage()));
-      return false;
-    }
-    return true;
+    return project.trySave(file);
   }
 
   @FXML

--- a/ui/src/main/java/edu/wpi/grip/ui/MainWindowController.java
+++ b/ui/src/main/java/edu/wpi/grip/ui/MainWindowController.java
@@ -1,5 +1,6 @@
 package edu.wpi.grip.ui;
 
+import edu.wpi.grip.core.GripFileManager;
 import edu.wpi.grip.core.Palette;
 import edu.wpi.grip.core.Pipeline;
 import edu.wpi.grip.core.PipelineRunner;
@@ -33,7 +34,6 @@ import java.io.File;
 import java.io.IOException;
 import java.util.Optional;
 import java.util.Set;
-import java.util.logging.Level;
 import java.util.logging.Logger;
 
 import javafx.application.Platform;
@@ -154,11 +154,7 @@ public class MainWindowController {
       } else if (dialog.getResult().equals(save)) {
         // If the user chose "Save", automatically show a save dialog and block until the user
         // has had a chance to save the project.
-        try {
-          return saveProject();
-        } catch (IOException e) {
-          logger.log(Level.SEVERE, e.getMessage(), e.getCause());
-        }
+        return saveProject();
       }
     }
     return true;
@@ -219,10 +215,18 @@ public class MainWindowController {
    * @return true if the user does not cancel the save
    */
   @FXML
-  public boolean saveProject() throws IOException {
+  public boolean saveProject() {
     if (project.getFile().isPresent()) {
       // Immediately save the project to whatever file it was loaded from or last saved to.
-      project.save(project.getFile().get());
+      try {
+        project.save(project.getFile().get());
+      } catch (IOException e) {
+        eventBus.post(new WarningEvent("Could not save project",
+            "The project could not be saved to " + project.getFile().get().getAbsolutePath()
+                + ".\n\nCause: " + e.getMessage()
+        ));
+        return false;
+      }
       return true;
     } else {
       return saveProjectAs();
@@ -237,10 +241,11 @@ public class MainWindowController {
    * @return true if the user does not cancel the save
    */
   @FXML
-  public boolean saveProjectAs() throws IOException {
+  public boolean saveProjectAs() {
     final FileChooser fileChooser = new FileChooser();
     fileChooser.setTitle("Save Project As");
     fileChooser.getExtensionFilters().add(new ExtensionFilter("GRIP File", "*.grip"));
+    fileChooser.setInitialDirectory(GripFileManager.GRIP_DIRECTORY);
 
     project.getFile().ifPresent(file -> fileChooser.setInitialDirectory(file.getParentFile()));
 
@@ -249,7 +254,14 @@ public class MainWindowController {
       return false;
     }
 
-    project.save(file);
+    try {
+      project.save(file);
+    } catch (IOException e) {
+      eventBus.post(new WarningEvent("Could not save project",
+          "The project could not be saved to " + file.getAbsolutePath()
+              + ".\n\nCause: " + e.getMessage()));
+      return false;
+    }
     return true;
   }
 

